File: 2022-04/javascript/150_Evaluate_Reverse_Polish_Notation.js
/**
 * https://leetcode.com/problems/evaluate-reverse-polish-notation/
 */


const operators = new Map([
    ['+', (x, y) => x + y],
    ['*', (x, y) => x * y],
    ['-', (x, y) => y - x],
    // 这里其实有好几种取整的方法，toFixed, parseInt, Math.floor, Math.ceil, Math.trunc, 位操作
    // 总的来说 Math.trunc() 与按位运算效率最高
    // n | 0
    // n ^ 0
    // n << 0
    ['/', (x, y) => Math.trunc(y / x)],
])

/**
 * @param {string[]} tokens
 * @return {number}
 */
function evalRPN(tokens) {
    const stack = []
    for (let t of tokens) {
        if (operators.has(t)) {
            stack.push(operators.get(t)(stack.pop(), stack.pop()))
        } else {
            stack.push(Number(t))
        }
    }
    return stack.pop()
}

File: 2022-04/python/150_Evaluate_Reverse_Polish_Notation.py
# https://leetcode.com/problems/evaluate-reverse-polish-notation/

class Solution:
    def __init__(self):
        self.operators = {
            '+': lambda x, y: x + y,
            '*': lambda x, y: x + y,
            '-': lambda x, y: y - x,
            # 这里需要向 0 靠拢，如果使用 floor/ceil 就需要按情况处理（负数是不一样的）
            # int() 操作正好简单的粗暴的干掉小数位
            '/': lambda x, y: int(y / x),  # truncate to zero
        }

    def eval_RPN(self, tokens: List[str]) -> int
        stack = []
        for t in tokens:
            if t in self.operators:
                stack.append(self.operators[t](stack.pop(), stack.pop()))
            else:
                stack.append(int(t))
        return stack.pop()
